ouroboros-consensus/src/Ouroboros/Consensus/Util/OptNP.hs

```diff
@@ -209,3 +212,51 @@ combine = combineWith $ \case
     This1 x   -> x
     That1 y   -> y
     These1 {} -> error "combine: precondition violated"
+
+{-------------------------------------------------------------------------------
+  Concatenation and appending
+-------------------------------------------------------------------------------}
+
+-- | Concatenation of type-level lists
+type family Concat xs ys where
+  Concat '[] ys = ys
+  Concat (x:xs) ys = x ': Concat xs ys
+
+-- | Type level append
+type (:|>) (xs :: [k]) x = Concat xs '[x]
+
+optConcat :: OptNP empty1 f xs -> OptNP empty2 f ys -> OptNP (empty1 && empty2) f (Concat xs ys)
+optConcat OptNil y         = y
+optConcat (OptSkip xs) y   = OptSkip $ optConcat xs y
+optConcat (OptCons x xs) y = OptCons x $ optConcat xs y
+
+optAppend :: OptNP empty f xs -> f x -> OptNP 'False f (xs :|> x)
+optAppend xs x = optConcat xs $ OptCons x OptNil
+
+optAppendSkip ::
```

Comment:
TODO

Is this only used in the `protocolInfoCardano` function? These combinators are well written, but my initial preference would be to altogether avoid the additional type-level computations they involve (`Concat` and `:|>`, namely).

I have an idea for how (approx: slightly enrich the type of the `maxTxCapacityOverridess` argument of `shelleySharedBlockForging` etc), if this is indeed their only use case.

Reply:
We discussed this on the call. Seems likely to tidy it up, but we'd have to work out the details. Relatively low priority.